cmd/pbstatuser/main.go

```diff
@@ -19,21 +29,101 @@ import (
+// in one of three channels (AWS, Azure, GCP)
+// 2) Reads from AWS channel, performs availability check, has configurable throttling.
+// 3) Reads from Azure channel, performs availability check, has configurable throttling.
+// 4) Reads from GCP channel, performs availability check, has configurable throttling.
+
+// Beware there is no database connection, no job queue, most of what api/worker processes have
+// is not available because it should not be needed. The worker will simply use SDKs to check
+// statuses and send out results via kafka.Send function to Sources.
+// Also make sure that all goroutines are closed correctly on context cancel. A WaitGroup
+// would be a really good here: https://gist.github.com/fracasula/b579d52daf15426e58aa133d0340ccb0
+
+var (
+	chProvider = make(chan clients.StatusResponse)
+	chBatch    = make(chan clients.SourcesKafkaPayload)
+	wg         = &sync.WaitGroup{}
+)
+
 func processMessage(ctx context.Context, message *kafka.GenericMessage) {
-	// TODO: implement the source checking for AWS, Azure, GCP
-
-	// This needs to be done in 4 goroutines.
-	// 1) Reads messages, calls sources to get auth type and ARN and then enqueue the work
-	// in one of three channels (AWS, Azure, GCP)
-	// 2) Reads from AWS channel, performs availability check, has configurable throttling.
-	// 3) Reads from Azure channel, performs availability check, has configurable throttling.
-	// 4) Reads from GCP channel, performs availability check, has configurable throttling.
-
-	// Beware there is no database connection, no job queue, most of what api/worker processes have
-	// is not available because it should not be needed. The worker will simply use SDKs to check
-	// statuses and send out results via kafka.Send function to Sources.
-	// Also make sure that all goroutines are closed correctly on context cancel. A WaitGroup
-	// would be a really good here: https://gist.github.com/fracasula/b579d52daf15426e58aa133d0340ccb0
+	// ctx, _ = context.WithCancel(ctx)
+
+	// Get source id
+	asm, _ := kafka.NewAvailabilityStatusMessage(message)
+	sourceId := asm.SourceID
+
+	// Get sources client
+	sourcesClient, err := clients.GetSourcesClient(ctx)
+	if err != nil {
+		return
+	}
+
+	// Fetch authentication from Sources
+	authentication, err := sourcesClient.GetAuthentication(ctx, sourceId)
+	if err != nil {
+		if errors.Is(err, clients.NotFoundErr) {
+			return
+		}
+		return
+	}
+
+	switch authentication.ProviderType {
+	case models.ProviderTypeAWS:
+		go ec2.CheckSourceAvailability(ctx, authentication, sourceId, chProvider)
+	// TODO: Complete
+	case models.ProviderTypeAzure:
+		go azure.CheckSourceAvailability(ctx, authentication, chProvider)
+	// TODO: Complete
+	case models.ProviderTypeGCP:
+		go gcp.CheckSourceAvailability(ctx, authentication, chProvider)
+	case models.ProviderTypeNoop:
+	case models.ProviderTypeUnknown:
+		// TODO: Complete
+	}
+	go createBatchKafkaSourcesPayload(ctx)
+}
```

Comment:
This is not how a typical Go program should look like when dealing with multiple background workers (goroutines). This function is called for every received message, you do not want to spawn goroutines. You want to start them only once from the `main()` function.

Then use package-level channels to communicate between them - pass structs from this function (Kafka receiver and authorization finder) to one of the three channels (aws, azure, gcp) where three functions (running a `range` loop as goroutines) will pick them up, work on them and pass results to final channel where another function will pick the results and send them.